cpp/include/cuspatial/experimental/detail/is_point_in_polygon_kernel.cuh

```diff
@@ -0,0 +1,113 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <cuspatial/traits.hpp>
+
+namespace cuspatial {
+namespace detail {
+
+/**
+ * @brief Kernel to test if a point is inside a polygon.
+ *
+ * Implemented based on Eric Haines's crossings-multiply algorithm:
+ * See "Crossings test" section of http://erich.realtimerendering.com/ptinpoly/
+ * The improvement in addenda is also addopted to remove divisions in this kernel.
+ *
+ * TODO: the ultimate goal of refactoring this as independent function is to remove
+ * src/utility/point_in_polygon.cuh and its usage in quadtree_point_in_polygon.cu. It isn't
+ * possible today without further work to refactor quadtree_point_in_polygon into header only
+ * API.
+ */
+template <class Cart2d,
+          class OffsetType,
+          class OffsetIterator,
+          class Cart2dIt,
+          class OffsetItDiffType = typename std::iterator_traits<OffsetIterator>::difference_type,
+          class Cart2dItDiffType = typename std::iterator_traits<Cart2dIt>::difference_type>
+__device__ inline bool is_point_in_polygon(Cart2d const& test_point,
+                                           OffsetType poly_begin,
+                                           OffsetType poly_end,
+                                           OffsetIterator ring_offsets_first,
+                                           OffsetItDiffType const& num_rings,
+                                           Cart2dIt poly_points_first,
+                                           Cart2dItDiffType const& num_poly_points)
+{
+  using T = iterator_vec_base_type<Cart2dIt>;
+
+  bool point_is_within = false;
+  bool is_colinear     = false;
+  // for each ring
+  for (auto ring_idx = poly_begin; ring_idx < poly_end; ring_idx++) {
+    int32_t ring_idx_next = ring_idx + 1;
+    int32_t ring_begin    = ring_offsets_first[ring_idx];
+    int32_t ring_end =
+      (ring_idx_next < num_rings) ? ring_offsets_first[ring_idx_next] : num_poly_points;
+
+    Cart2d b     = poly_points_first[ring_end - 1];
+    bool y0_flag = b.y > test_point.y;
+    bool y1_flag;
+    // for each line segment, including the segment between the last and first vertex
+    for (auto point_idx = ring_begin; point_idx < ring_end; point_idx++) {
+      Cart2d const a = poly_points_first[point_idx];
+      T run          = b.x - a.x;
+      T rise         = b.y - a.y;
+
+      // Points on the line segment are the same, so intersection is impossible.
+      // This is possible because we allow closed or unclosed polygons.
+      if (run == 0.0 && rise == 0.0) continue;
+
+      T rise_to_point = test_point.y - a.y;
+
+      // colinearity test
+      const T EPSILON = 0.0000000000000000000000000000001;
```

Comment:
The one last thing I have with is that the epsilon value should be machine/precision dependent. Maybe a multiples of `numeric_limit<T>::epsilon()` or multiples of ULP.

Reply:
Unscaled, `numeric_limits<T>::epsilon()` is by definition 1 ULP, isn't it? So should probably be like `k * numeric_limits<T>::epsilon()` for some small k.  Normally you would scale by the magnitude of the value being compared, but since that value is assumed to be close to zero, there's no need.

Reply:
> Unscaled, numeric_limits<T>::epsilon() is by definition 1 ULP, isn't it? 
Yes. Suggest `k=4` to align with gtest?

Reply:
In a long run, I was hoping to expose the precision issue as a user configurable option. As this issue will soon come up else as we encounters more complicated ops.

Reply:
File an issue for that, please.   We may want a way to specify global settings like this, if we don't want to add a precision argument to every API...